### DISCUSSION TITLE
Can I build a mnemonic menu using keyboard shortcuts consisting of single letter keystrokes (without modifiers)?

### DISCUSSION THREAD

**Discussion by stephenkraemer** [microsoft/vscode-discussions#2417] [Extension Development QnA]
I would like to add a spacemacs-like mnemonic menu to my extension.

For this purpose, I would be grateful for a clarification whether "single-letter" keyboard shortcuts are supported. For example, it is currently in principle possible to define a keyboard shortcut: "ctrl+k d p" (which triggers when pressing ctrl+k, then d, then p). It is not nessary to use a modifier with "d" or "p" (which would be the "emacs-style" keyboard shortcut "ctrl+k ctrl+d ctrl+p).

However
- I see nothing in the documentation that indicates that such "single-letter" keybindings are explicitely supported (or explicitely not supported).
- There is as far as i can see not a single keybinding in the vscode defaults which uses this capability. Instead all keybindings use modifiers for each key press (ctrl+k ctrl+d ctrl+p).
- Related extensions like edamagit implement menus using single-letter keyboard shortcuts in combination with specific when-clauses. But I have not seen keyboard shortcuts defining a sequence of single letter presses.

I am aware that VSpaceCode has implemented mnemonic menus using vscode-which-key, but this is overkill for my purposes and adds an external dependency that I would like to avoid if possible. My extension has no need for a graphical display / search of the available keyboard shortcuts, which is the main advantage of vscode-which-key as far as I can see.

Is the "single-letter" style ("ctrl+k d p") reliably supported? Is there a reason why it appears to be rarely used in extensions or the vscode defaults?

Motivation: such single-letter style keyboard shortcuts would allow extension authors to define a spacemacs-like mnemonic menu, giving a modal editing experience in vscode (that goes easier on the hands than the classic emacs style of ctrl+k ctrl+d ctrl+p). Going further, this could even be combined with extension-specific when-clauses to implement quite powerful modal editing.

Thank you!

**Comment by isidorn** [comment]
I am not 100% sure. Can you please check in the docs https://code.visualstudio.com/docs/getstarted/keybindings

Also can you just try it out in your extensions, and if things do not work you could share some code and I can double check with our dev team. Thanks

**Reply by stephenkraemer** [reply]
Thank you very much for your reply!

I have tried it out in different ways. I works very well! My problem is that I am not sure if this is actually supported or it just works by chance and will stop working in the future.

 I think the relevant section from the documentation is here: 

https://code.visualstudio.com/docs/getstarted/keybindings#_accepted-keys

where it states

> The key is made up of modifiers and the key itself. The following modifers are accepted ... {Ctrl, Alt etc.}. The following keys are accepted ... {basically any key}.

To me, the documentation suggests that any key must be combined with a modifier, but I think it isn't 100% clear on that. That's why I simply tried my one-letter keybindings and found that there is really no limit to their usage. This is really powerful, and I hope that others could benefit from "single-letter" keybindings too.

As a simple example, these work out of the box when placed  in my package.json or the user keyboard shortcuts. I am using a vscode command to make this reproducible:

```
code example 1:
      {
        "key": "Ctrl+k b r",
        "command": "workbench.action.files.revert",
      }
```

Here I use a modifier to "escape from insert mode", and afterwords I can use single key combos like "b, then r". If I press an unexpected key (e..g "a") nothing happens, but I am "back in insert mode".  If I complete any of the defined sequences, the command is triggered, and I am back in insert mode.

I have found that I can go even further. My extension defines a custom when clause context "vscode-modal.mode". I set the mode like this:

```
vscode.commands.executeCommand('setContext', 'vscode-modal.mode', 'command');
```

This allows me to define a space-based menu like in spacemacs and co:

```
code example 2
      {
        "key": "space b r",
        "command": "workbench.action.files.revert",
        "when": "vscode-modal.mode == 'normal' && (editorTextFocus || notebookEditorFocused)"
      }
```

So it works and is really powerful. It basically allows turning vscode into a fully modal editor with minimal effort. But I haven't found any default keybinding using this, nor any other extension using this, and I don't fully understand the wording of the documentation.

Many extension jump through a lot of hoops to get such modal behavior, e.g. https://github.com/DCsunset/vscode-modal-editor (I am aware that the extension does a lot of things beyond simple modal editing, like recording past commands, I just want to use this as a general example)

I believe that a clarification in the docs on what is directly supported by keyboard shortcuts would be helpful for many people interested in modal editing!

To summarize my questions are:
1. Are "one-letter" keyboard shortcuts supported? (Code example 1)
2. Is creating modal menus with "one-letter" keyboard shortcuts and setContext to create custom when-clauses supported? (Code example 2). Of course one will always need to enter the "normal mode" through a standard, modifier-based keybinding.
3. Are there drawbacks to this approach, e.g. with regard to performance?

Thank you!

**Reply by stephenkraemer** [reply]
Hello and happy new year! I realize that over Christmas it is highly unlikely that anybody found time for this question, which is of course perfectly alright. I was wondering whether you could have a look at it again sometime in the next weeks? Many thanks in advance!

**Reply by isidorn** [reply]
@hediet would be best to answer your 3 questions.
But he is on vacation this month. So his response might be delayed.

**Reply by stephenkraemer** [reply]
Thanks again for your help! I am happy to wait of course!

**Comment by ArturoDent** [comment]
`"key": "ctrl+k b r",` is, I believe, perfectly legal and supported (it isn't a single letter keybinding but a keychord), but there may be an issue for you.  

After you type `ctrl+k` it waits for the next key, but it will timeout in a relatively short period of time, after which it is back in insert mode.
